src/apiutil.h

```diff
@@ -308,3 +316,208 @@ bool hasInsufficientMaterial(Color c, const Position& pos) {
+    if (nbProcessedRanks != nbRanks) {
+      std::cerr << "Invalid number of ranks. Expected: " << nbRanks << " Actual: " << nbProcessedRanks << std::endl;
+      return FEN_INVALID_NUMBER_OF_RANKS;
+    }
+  }
+
+  // check for pocket
+  if (v->pieceDrops) {
+    if (nbProcessedRanks-1 == nbRanks) {
+      // look for last '/'
+      for (auto it = fenParts[0].rbegin(); it != fenParts[0].rend(); ++it) {
+        if (*it == '/')
+          break;
+        if (*it != '-') {
+          if (v->pieceToChar.find(*it) == std::string::npos) {
+            std::cerr << "Invalid pocket piece: '" << *it << "'." << std::endl;
+            return FEN_INVALID_POCKET_INFO;
+          }
+        }
+      }
+    }
+    else {  // pocket is defined as [ and ]
+      if (*(fenParts[0].end()-1) != ']') {
+        std::cerr << "Pocket specification does not end with ']'." << std::endl;
+        return FEN_INVALID_POCKET_INFO;
+      }
+      // look for first '['
+      for (auto it = fenParts[0].rbegin()+1; it != fenParts[0].rend(); ++it) {
+        if (*it == '[')
+          break;
+        if (*it != '-')
+          if (v->pieceToChar.find(*it) == std::string::npos) {
+              std::cerr << "Invalid pocket piece: '" << *it << "'." << std::endl;
+            return FEN_INVALID_POCKET_INFO;
+          }
+      }
+    }
+  }
+
+  // 2) Part
+  // check side to move char
+  if (fenParts[1][0] != 'w' && fenParts[1][0] != 'b') {
+    std::cerr << "Invalid side to move specification: '" << fenParts[1][0] << "'." << std::endl;
+    return FEN_INVALID_SIDE_TO_MOVE;
+  }
+
+  // 3) Part
+  // check castling rights (exclude seirawan_variant because it has different castling specification)
+  if (v->castling && !v->seirawanGating) {
+    for (char c : fenParts[2])
+      if (c != '-')
+        if (c != 'k' && c != 'K' && c != 'Q' && c != 'q') {
+          std::cerr << "Invalid castling specification: '" << c << "'." << std::endl;
+          return FEN_INVALID_CASTLING_INFO;
+        }
+  }
+
+  // 4) Part
+  // check en-passant square
+  if (v->castling) {
+    const char firstChar = fenParts[3][0];
+    if (firstChar != '-') {
+      if (fenParts[3].size() != 2) {
+        std::cerr << "Invalid en-passant square. Expects 2 characters. Actual: '" << fenParts[3].size() << " characters." << std::endl;
+        return FEN_INVALID_EN_PASSANT_SQ;
+      }
+      if (isdigit(firstChar)) {
+        std::cerr << "Invalid en-passant square. Expects 1st character to be a digit." << std::endl;
+        return FEN_INVALID_EN_PASSANT_SQ;
+      }
+      const char secondChar = fenParts[3][1];
+      if (!isdigit(secondChar)) {
+        std::cerr << "Invalid en-passant square. Expects 2nd character to be a non-digit." << std::endl;
+        return FEN_INVALID_EN_PASSANT_SQ;
+      }
+    }
+  }
+
+  // 5) Part
+  // check checkCounting (the karouk_variant doesn't feature a checkCounter within the FEN because the first check wins)
+  if (v->checkCounting && !v->cambodianMoves)
```

Comment:
Whether the checkCounting field is skipped is not related to cambodian chess, but only to whether the counts are exactly 1. This is a bit ugly from the validation perspective, but it basically means that skipping this field is always allowed, it just implies that the counts are exactly 1. I would also be fine with allowing skipping only for cambodian chess (since allowing it is kind of arbitrary anyway and just related to GUI implementations), but the scenario of cambodian chess with a count > 1 should be supported, i.e., the field should still be allowed to be present as otherwise there are hypothetical variants where no FEN would be valid.

Reply:
I'll remove this check then to allow skipping the `checkCounter` in cases where only check is required.